timemachine/cpp/src/kernels/k_exchange.cu

```diff
@@ -308,54 +331,65 @@ template void __global__ k_compute_box_volume<double>(
 template <typename RealType, bool Negated>
 void __global__ k_adjust_weights(
     const int N,
-    const int num_target_mols,
+    const int batch_size,
     const int mol_size,
+    const int num_weights,
     const int *__restrict__ mol_atoms_idxs,
     const int *__restrict__ mol_offsets,
     const RealType *__restrict__ per_atom_energies,
     const RealType inv_kT, // 1 / kT
     RealType *__restrict__ log_weights) {
-    int mol_idx = blockIdx.x * blockDim.x + threadIdx.x;
-    while (mol_idx < num_target_mols) {
 
-        RealType current_log_weight = log_weights[mol_idx];
-        __int128 weight_accumulator = 0;
+    int idx_in_batch = blockIdx.y;
+    while (idx_in_batch < batch_size) {
 
-        int mol_start = mol_offsets[mol_idx];
-        int mol_end = mol_offsets[mol_idx + 1];
-        int min_atom_idx = mol_atoms_idxs[mol_start];
-        int max_atom_idx = mol_atoms_idxs[mol_end - 1];
+        int mol_idx = blockIdx.x * blockDim.x + threadIdx.x;
+        while (mol_idx < num_weights) {
+
+            RealType current_log_weight = log_weights[idx_in_batch * num_weights + mol_idx];
+            __int128 weight_accumulator = 0;
+
+            int mol_start = mol_offsets[mol_idx];
+            int mol_end = mol_offsets[mol_idx + 1];
+            int min_atom_idx = mol_atoms_idxs[mol_start];
+            int max_atom_idx = mol_atoms_idxs[mol_end - 1];
 
-        // A loop that in the case of water will be 3x3
-        for (int i = 0; i < mol_size; i++) {
-            for (int j = min_atom_idx; j <= max_atom_idx; j++) {
-                weight_accumulator += FLOAT_TO_FIXED_ENERGY<RealType>(inv_kT * per_atom_energies[i * N + j]);
+            // A loop that in the case of water will be 3x3
+            for (int i = idx_in_batch * mol_size; i < (idx_in_batch + 1) * mol_size; i++) {
+                for (int j = min_atom_idx; j <= max_atom_idx; j++) {
+                    weight_accumulator += FLOAT_TO_FIXED_ENERGY<RealType>(inv_kT * per_atom_energies[i * N + j]);
+                }
             }
-        }
 
-        weight_accumulator = Negated ? FLOAT_TO_FIXED_ENERGY<RealType>(current_log_weight) - weight_accumulator
-                                     : FLOAT_TO_FIXED_ENERGY<RealType>(current_log_weight) + weight_accumulator;
+            weight_accumulator = Negated ? FLOAT_TO_FIXED_ENERGY<RealType>(current_log_weight) - weight_accumulator
+                                         : FLOAT_TO_FIXED_ENERGY<RealType>(current_log_weight) + weight_accumulator;
 
-        log_weights[mol_idx] =
-            fixed_point_overflow(weight_accumulator) ? INFINITY : FIXED_ENERGY_TO_FLOAT<RealType>(weight_accumulator);
+            log_weights[idx_in_batch * num_weights + mol_idx] =
```

Comment:
sanity check that every thread writes to a unique address for all possible values of `num_weights` and `batch_size`?

```
idx_in_batch * num_weights + mol_idx
=
(blockIdx.y+b*gridDim.x * blockDim.x)*num_weights + blockIdx.x * blockDim.x + threadIdx.x + w*(gridDim.x * blockDim.x) for integer b, w values
```

Reply:
Sanity checked.

Breaks down to `[0, K) * N + [0, N)` where N is the number of weights and K is the batch size